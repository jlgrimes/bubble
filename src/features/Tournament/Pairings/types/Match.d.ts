import { MatchResult } from "./MatchResult"

export interface Match {
  /**
   * Result of the match. First player against second player (ex, win means player 1 beat player 2)
   */
  result: MatchResult,
  /**
   * Player ids of the paired players
   */
  playerIds: number[]
}