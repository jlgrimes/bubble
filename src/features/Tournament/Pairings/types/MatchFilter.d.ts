export type MatchFilter = 'completed' | 'incomplete' | undefined;