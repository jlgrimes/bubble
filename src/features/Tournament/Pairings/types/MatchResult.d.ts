export type MatchResult = 'win' | 'loss' | 'tie'