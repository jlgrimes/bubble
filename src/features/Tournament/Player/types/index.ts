import { Player } from "./Player";
import { PlayerMatch } from "./PlayerMatch";
import { Record } from "./Record";

export type { Player, PlayerMatch, Record };