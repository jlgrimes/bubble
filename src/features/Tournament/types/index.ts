import { TournamentState } from "./Tournament";

export type { TournamentState };