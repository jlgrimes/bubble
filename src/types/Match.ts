export enum Result {
  Win,
  Loss,
  Tie
}

export interface Match {
  /**
   * Player ids of the paired players
   */
  playerIds: String[],
  /**
   * Result of the match. First player against second player (ex, win means player 1 beat player 2)
   */
  result?: Result
}