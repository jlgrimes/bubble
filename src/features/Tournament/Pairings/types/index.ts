import { Match } from "./Match";
import { MatchResult } from "./MatchResult";

export type { Match, MatchResult };