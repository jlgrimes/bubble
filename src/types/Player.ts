import { Match, PlayerMatch } from "./Match";
import { Record } from "./Record";

export interface Player {
  /**
   * Unique identifier of the player
   */
  id: string;
  /**
   * Name of the player. Ex. Kenny
   */
  name: string;
  /**
   * Matches the player has played. Starts as empty array
   */
  matches: PlayerMatch[];
  /**
   * The record a player has. Should be updated alongside matches.
   */
  record: Record;
  /**
   * Dropped status. If true, player is dropped, do not include in pairings.
   */
  dropped?: boolean;
};