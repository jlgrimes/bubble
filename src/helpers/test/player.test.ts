import { Match, PlayerMatch } from '../../types/Match';
import { Player } from '../../types/Player';
import {
  convertMatchToPlayerMatch,
  getUpdatedPlayerAfterMatch,
  chunkSortedArrayByMatchPoints,
  shuffleAndFlattenChunkedPlayersByMatchPoints,
  sortPlayersByMatchPoints,
} from '../player';
import { shuffle } from '../shuffle';
import {
  JARED,
  RYAN,
  KENNY,
  NOAH,
  SAMPLE_PLAYER,
  SAMPLE_SORTED_PLAYER_LIST,
} from './shared';

describe('player helpers', () => {
  describe('convertMatchToPlayerMatch', () => {
    it('should preserve match result if player is listed first', () => {
      const match: Match = { result: 'win', playerIds: ['0', '4'] };
      const expectedPlayerMatch: PlayerMatch = {
        opponentId: '4',
        result: 'win',
      };

      expect(convertMatchToPlayerMatch(SAMPLE_PLAYER, match)).toEqual(
        expectedPlayerMatch
      );
    });

    it('should inverse match result if player is listed second', () => {
      const match: Match = { result: 'win', playerIds: ['4', '0'] };
      const expectedPlayerMatch: PlayerMatch = {
        opponentId: '4',
        result: 'loss',
      };

      expect(convertMatchToPlayerMatch(SAMPLE_PLAYER, match)).toEqual(
        expectedPlayerMatch
      );
    });
  });

  describe('getUpdatedPlayerAfterMatch', () => {
    it('should get correct player after match', () => {
      const match: Match = { result: 'win', playerIds: ['0', '4'] };
      const expectedPlayerMatch: PlayerMatch = {
        opponentId: '4',
        result: 'win',
      };
      const expectedPlayer: Player = {
        ...SAMPLE_PLAYER,
        matches: [...SAMPLE_PLAYER.matches, expectedPlayerMatch],
        record: {
          wins: 2,
          ties: 1,
          losses: 1,
        },
        matchPoints: 7,
      };

      expect(getUpdatedPlayerAfterMatch(SAMPLE_PLAYER, match)).toEqual(
        expectedPlayer
      );
    });
  });

  describe('chunkSortedArrayByMatchPoints', () => {
    it('should return empty array if no players', () => {
      expect(chunkSortedArrayByMatchPoints([])).toEqual([]);
    });

    it('should chunk array by match points', () => {
      expect(
        chunkSortedArrayByMatchPoints(SAMPLE_SORTED_PLAYER_LIST)
      ).toMatchSnapshot();
    });
  });

  describe('shuffleAndFlattenChunkedPlayersByMatchPoints', () => {
    it('should shuffle and flatten chunkedPlayerIdxs', () => {
      const chunk = [[0], [1, 2], [3]];
      expect([
        [0, 1, 2, 3],
        [0, 2, 1, 3],
      ]).toContainEqual(shuffleAndFlattenChunkedPlayersByMatchPoints(chunk));
    });
  });

  describe('sortPlayersByMatchPoints', () => {
    it('should sort and shuffle players by their match points', () => {
      const scrambledList = shuffle(SAMPLE_SORTED_PLAYER_LIST);
      expect([
        [JARED, NOAH, RYAN, KENNY],
        [JARED, RYAN, NOAH, KENNY]
      ]).toContainEqual(
        sortPlayersByMatchPoints(scrambledList)
      );
    });
  });
});
