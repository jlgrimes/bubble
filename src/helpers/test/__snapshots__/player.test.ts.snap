// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`player helpers chunkSortedArrayByMatchPoints should chunk array by match points 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    1,
    2,
  ],
  Array [
    3,
  ],
]
`;
